### DISCUSSION TITLE
Weird behaviour with raycasted setup & extern systems

### DISCUSSION THREAD

**Discussion by frutas-fruit** [xissburg/edyn#102] [General]

**Comment by frutas-fruit** [comment]


**Reply by xissburg** [reply]
So this is as if no `RaycastExtern` component is present in the registry?

**Reply by xissburg** [reply]
Your setup looks fine to me. It just looks like the ray entity is not being transported when islands are merged. Either that or the `RaycastExtern` component is not being moved into the other island. Can you try printing the entities present in a `m_registry->view<edyn::null_constraint>`? That will show us whether the ray entities are being migrated.

**Reply by xissburg** [reply]
> I've also tried enabling just 1 point constraint (connecting the head to torso), and my player goes crazy when I try and apply a grab force to his torso:

Could it be because it has too little mass?

**Comment by frutas-fruit** [comment]


**Reply by xissburg** [reply]
The `null_constraint` creates an edge between two entities in the entity graph, ensuring they're always in the same island. Otherwise they're only gonna be in the same island if they're about to collide. You definitely do not want to have that between the vehicle and ragdoll, but it's ok to try it and see what happens.

The order with which rigid bodies are replicated in the background thread should not be relevant, and I don't think the order is gonna be exactly the same as the order they're created in the main thread, since they're collected and dispatched to islands in background threads by traversing the entity graph.

**Reply by xissburg** [reply]
Yeah it looks fine. You register the external components and external system once and that's all. Can you confirm that the external system function is being called? Or is it just the view that's empty?

**Reply by xissburg** [reply]
So it looks like as soon as the ragdoll moves into the vehicle's island, the raycasts start to fail. So it just returns a null entity? Can't you step into the function call with a debugger?

**Comment by xissburg** [comment]
I understand what is happening here, but I am not sure of the reason. The vehicle and ragdoll are initially in separate islands, running in different threads. When they get close to one another, their AABB intersect and they're moved into a single island. In the case where you have multiple ragdolls, whose AABBs are apparently intersecting (thus meaning they're in the same island), when the car approaches one ragdoll, it is transported into that same island and the island it was in is destroyed. Remember that each island has their own private registry. The process of moving an entity from one island to another involves making a copy of all of its components in the local registry and sending it over to the other registry, that's why you need to register the external components explicitly or else it wouldn't know what components to copy. What must be happening is that the external components are not being copied so they're lost and the view you're iterating in the external system is empty.

I don't understand why they're not being copied. I will have to investigate if any recent changes I made broke anything.

All the weird behavior you're seeing is being caused by this "small" failure. I get it that it can be difficult to understand what is happening without a deeper knowledge about how the multi threaded physics is setup, where any slight mistake can lead to situations where the behavior cannot be explained by physics alone.

**Comment by frutas-fruit** [comment]


**Reply by xissburg** [reply]
Creating a `null_constraint` shouldn't be necessary, because as soon as their AABBs intersect, a `contact_constraint` is created between the vehicle and ragdoll, which is an edge in the entity graph, which consequently merges their islands into one, so it has the same effect.

The raycast is executed on the island's broadphase tree, so all entities in that island are considered. It doesn't look like the rays are intersecting the ragdoll btw in the earlier videos, so... I am wondering now whether there's a bug in the raycast function specifically when called in an external system, which calls into the `broadphase_worker`, whereas when raycasting in the main thread it calls into the `broadphase_main`. I will have to conduct a few tests.

**Comment by xissburg** [comment]
I have discovered the reason behind this issue with the vehicle suddenly dropping to the floor. 

One issue is that the ray was intersecting the car itself so it was basically applying the same force in opposite directions to the car (thus nullifying it) and then nothing happens. To fix that I have added a filter to the `edyn::raycast` function which can be passed in the last parameter and return true if the entity hit by the ray is the car.

The other issue is that the AABB of the vehicle was not necessarily intersecting the AABB of the terrain or plane or ragdoll since it's floating due to the ray forces. That would cause the entities to not be placed in the same island then raycasts would be missed because raycasts in an external system only consider the entities that are in the island where it's running at the given moment. To fix that I have added a new componnet `edyn::AABB_override` which allows the AABB of the shape to be overridden by a custom AABB which encloses the shape and the rays. That way entities will be moved into the same island even if their actual AABBs are not intersecting yet and the rays will hit the nearest entity as expected. I've written a bit about it in the wiki https://github.com/xissburg/edyn/wiki/User-Manual#aabb-override

Here you can find an example of a hovering rigid body which is similar to what you're doing with a raycast vehicle https://github.com/xissburg/edyn-testbed/blob/master/client/src/hover.cpp

Hope this helps. Cheers!
